src/boot/efi/devicetree.c

```diff
@@ -104,6 +106,65 @@ EFI_STATUS devicetree_install(struct devicetree_state *state, EFI_FILE *root_dir
                         MAKE_GUID_PTR(EFI_DTB_TABLE), PHYSICAL_ADDRESS_TO_POINTER(state->addr));
 }
 
+static char* devicetree_get_compatible(const void *dtb) {
+        const struct fdt_header *dt_header = dtb;
+        uint32_t *cursor = (uint32_t *) ((char *)dt_header + be32toh(dt_header->off_dt_struct));
+        char *string_block = (char *) ((char *)dt_header + be32toh(dt_header->off_dt_strings));
+        uint32_t len;
+
+        if (be32toh(dt_header->magic) != 0xd00dfeed)
+                return NULL;
+
+        while (true) {
+                switch (be32toh(*cursor++)) {
+                case FDT_BEGIN_NODE:
+                        if (*cursor++ != 0)
+                                return NULL;
+                        break;
+                case FDT_NOP:
+                        break;
+                case FDT_PROP:
+                        len = be32toh(*cursor++);
+
+                        if (strcmp8(string_block + be32toh(*cursor++), "compatible") == 0)
+                                return (char *)cursor;
+
+                        cursor += (len + 3) / 4;
```

Comment:
What are these numbers?

Reply:
These numbers are for rounding up the cursor to the next word, since len can be any integer. However I can change how this is done if this is not so clear to see when looking at the code